### ISSUE TITLE
Pass Babel AST directly

### ISSUE THREAD

**Issue by michael-ciniawsky** [babel/babel-loader#539]
**I'm submitting a feature request**

**Webpack Version:**

`v4.0.0 (@next)`

**Babel Core Version**:

`v7.0.0 (@next)`

**Babel Loader Version**:

`v8.0.0`

**Current behavior:**

Pass `result.code` `{String}`

**Desired behavior:**

Pass `result.ast` as metadata `cb(null, code, map, { webpackAST: ast /* result.ast */ })` `{Object}`

> ⚠️  Needs to be compatible to the Acorn AST (ESTree) ([webpack #5925](https://github.com/webpack/webpack/pull/5925))

* **What is the motivation / use case for changing the behavior?**

Avoid unnecessary parsing, if possible

**Comment by xtuc** [comment]
I started to work on this.

The only change in babel-loader is to emit an AST instead of the source.

We need to convert our Babel AST to the ESTree, which shouldn't be too hard.

**Comment by hzoo** [comment]
Depends on how it works, may need it to be transformed both ways if Babel loader is passed an ast itself, easiest thing to do to output the ast would be if we wrote a Babel transform for it but not sure we can handle that with different nodes atm

**Comment by xtuc** [comment]
I assumed that Babel would parse and apply transformations. But it seems that works only that way? I don't see where the loader has access to the ast?

**Comment by michael-ciniawsky** [comment]
It doesn't have access atm, 

```diff
- function loader (src, map)
+ function loader (src, map, { webpackAST })
```

But if there needs to be a Tree Adapter of some sort the question is if this isn't actually slower then letting webpack parse it again :)

**Comment by xtuc** [comment]
> But if there needs to be a Tree Adapter of some sort the question is if this isn't actually slower then letting webpack parse it again :)

I'm sure it will be faster. Parsing text is way slower than juste mutations on the AST and given that the ASTs aren't so much different.

**Comment by hzoo** [comment]
https://twitter.com/loganfsmyth/status/929564354571157504

**Comment by xtuc** [comment]
It seems that a converter already exists https://www.npmjs.com/package/babel-to-estree

**Comment by hzoo** [comment]
> We have the estree logic in Babylon, babel-eslint, espree

It is a fork of what we already had in babel-eslint and both might be out of date

**Comment by loganfsmyth** [comment]
In recently playing around with caching in babel-core, I've noticed that my test logic (so I'd assume `babel-loader` too) spends a decent chunk of time in JSON.stringify and JSON.parse on Babel's result object because of the AST. I wonder how performance gains from `webpackAST` would compare to the gains of passing `ast: false` to Babel, so it doesn't have to cache the AST as JSON, but then also wouldn't have it to pass off to Webpack. Webpack would have to parse Babel's output string, but it'd mean we don't have to re-parse the AST from JSON. While simpler to parse, the JSON is dramatically larger. Not sure where that tradeoff would be.

**Comment by DrewML** [comment]
I have a few different POCs of this locally with `webpack@next` and various different ways of going from Babylon >> ESTree. Just looking for some medium/large sized projects to profile on to determine if this is worthwhile to bring into the project.

**Comment by loganfsmyth** [comment]
I'll just register that if we do find we want to land this, I think I'd push for it being opt-in. When caching lands in core, it'll mean we just have to spend tons of time parsing JSON instead of tons of time parsing JS, and I'm pretty sure the AST JSON is so large that it takes longer to parse than the actual JS content would take.

I could see this being a performance boost if the user isn't using any caching, but I'm not sure that's a case we should bother optimizing for?

**Comment by DrewML** [comment]
> I think I'd push for it being opt-in.

Agreed. I'm still _highly_ skeptical this will have any real perf benefits, but I wanted to actually profile it so we can make a decision, document it, and put it to rest 😄

**Comment by DrewML** [comment]
Looks like this is more challenging that just passing an ESTree-compatible AST back to `webpack`.

I created the [lazy-babylon-to-estree](https://github.com/DrewML/lazy-babylon-to-estree/) package to do the minimum amount of changes needed for `webpack` to be happy.

After running tests on some larger projects, I hit a major blocker: anytime a transform creates a new node, that new node won't have a `start`/`end`/`range`. Webpack relies on those values existing, since it does string-replacement on sources rather than a full print from the AST.

**Example**
`<Foo />`is going to transform to `React.createElement(Foo, null)`. `webpack` expects that member expression to have a range because, when the `React` import declaration is swapped, `webpack` needs to replace usages of the `React` identifier.

**Comment by loganfsmyth** [comment]
> After running tests on some larger projects, I hit a major blocker: anytime a transform creates a new node, that new node won't have a start/end/range.

Great point, and thinking on that even more, it's not even just that. The location ranges in the AST reference the locations in the _original_ source content, not the position in the resulting code string.

**Comment by DrewML** [comment]
> The location ranges in the AST reference the locations in the original source content passed to the loader, not the position in the resulting code string returned from the loader.

Oof, I hadn't even considered that. I hate to say it, but I think this idea is pretty much dead in the water.

**Comment by jameswomack** [comment]
I'd love to have an opt-in version of this. It'd enable me to process/act on call expressions (or other fragments) that the Webpack AST hooks aren't able to reach, including:
* Those in JSX
* Those in `if` statements, `switch` statements deeper in module hierarchies etc.

**Comment by reuwi** [comment]
It's been 5 years since this issue was created...so has any progress?
